File: internal/tools/notes/notes.go
package notes

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type NoteTool struct {
	dataPath string
}

func NewNotesTool() *NoteTool {
	workingDir, err := os.Getwd()
	if err != nil {
		log.Printf("Error getting working directory: %v\n", err)
		return nil
	}

	dataDir := filepath.Join(workingDir, "data", "notes")
	if err := os.MkdirAll(dataDir, 0755); err != nil {
		log.Printf("Error creating notes directory: %v\n", err)
		return nil
	}

	return &NoteTool{
		dataPath: dataDir,
	}
}

type NoteArguments struct {
	Action    string `json:"action"`
	Title     string `json:"title"`
	Content   string `json:"content"`
	Search    string `json:"search"`
	StartDate string `json:"start_date"`
	EndDate   string `json:"end_date"`
}

type Note struct {
	Title     string    `json:"title"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (n *NoteTool) CallTool(arguments string) string {
	var args map[string]interface{}
	if err := json.Unmarshal([]byte(arguments), &args); err != nil {
		return fmt.Sprintf("Error parsing arguments: %v", err)
	}

	userID, ok := args["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	userDataPath := filepath.Join(n.dataPath, userID)
	if err := os.MkdirAll(userDataPath, 0755); err != nil {
		return fmt.Sprintf("Error creating user notes directory: %v", err)
	}

	var noteArgs NoteArguments
	if b, err := json.Marshal(args); err == nil {
		_ = json.Unmarshal(b, &noteArgs)
	}

	if err := n.validateInput(noteArgs); err != nil {
		return fmt.Sprintf("Validation error: %v", err)
	}

	switch strings.ToUpper(noteArgs.Action) {
	case "GET":
		return n.getNotes(userDataPath)
	case "GET_DETAIL":
		return n.getNoteDetail(userDataPath, noteArgs.Title)
	case "POST":
		return n.saveNote(userDataPath, noteArgs.Title, noteArgs.Content)
	case "PUT":
		return n.updateNote(userDataPath, noteArgs.Title, noteArgs.Content)
	case "DELETE":
		return n.deleteNote(userDataPath, noteArgs.Title)
	case "SEARCH":
		return n.searchNotes(userDataPath, noteArgs.Search)
	case "GET_BY_DATE":
		return n.getNotesByDate(userDataPath, noteArgs.StartDate, noteArgs.EndDate)
	default:
		return "Invalid action specified. Please use GET, GET_DETAIL, POST, PUT, DELETE, SEARCH, or GET_BY_DATE."
	}
}

func (n *NoteTool) validateInput(args NoteArguments) error {
	if args.Action == "" {
		return fmt.Errorf("action is required")
	}

	if args.Action == "POST" || args.Action == "PUT" {
		if args.Title == "" {
			return fmt.Errorf("title is required")
		}
		if args.Content == "" {
			return fmt.Errorf("content is required")
		}
	}

	return nil
}

func (n *NoteTool) getNotes(userDataPath string) string {
	files, err := os.ReadDir(userDataPath)
	if err != nil {
		return fmt.Sprintf("Error reading notes directory: %v", err)
	}

	var notes []Note
	for _, file := range files {
		if !file.IsDir() {
			note, err := n.readNoteFile(userDataPath, file.Name())
			if err != nil {
				continue
			}
			notes = append(notes, note)
		}
	}

	jsonNotes, err := json.Marshal(notes)
	if err != nil {
		return fmt.Sprintf("Error marshaling notes: %v", err)
	}

	return string(jsonNotes)
}

func (n *NoteTool) readNoteFile(userDataPath, filename string) (Note, error) {
	filePath := filepath.Join(userDataPath, filename)
	content, err := os.ReadFile(filePath)
	if err != nil {
		return Note{}, err
	}

	var note Note
	if err := json.Unmarshal(content, &note); err != nil {
		return Note{}, err
	}

	return note, nil
}

func (n *NoteTool) getNoteDetail(userDataPath, title string) string {
	note, err := n.readNoteFile(userDataPath, fmt.Sprintf("%s.json", title))
	if err != nil {
		return fmt.Sprintf("Error reading note %s: %v", title, err)
	}

	jsonNote, err := json.Marshal(note)
	if err != nil {
		return fmt.Sprintf("Error marshaling note: %v", err)
	}

	return string(jsonNote)
}

func (n *NoteTool) saveNote(userDataPath, title, content string) string {
	filePath := filepath.Join(userDataPath, fmt.Sprintf("%s.json", title))

	if _, err := os.Stat(filePath); err == nil {
		return fmt.Sprintf("Note with title '%s' already exists. Use PUT to update it.", title)
	}

	note := Note{
		Title:     title,
		Content:   content,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	jsonNote, err := json.Marshal(note)
	if err != nil {
		return fmt.Sprintf("Error marshaling note: %v", err)
	}

	if err := os.WriteFile(filePath, jsonNote, 0644); err != nil {
		return fmt.Sprintf("Error saving note: %v", err)
	}

	return fmt.Sprintf("Note '%s' has been saved successfully.", title)
}

func (n *NoteTool) updateNote(userDataPath, title, content string) string {
	filePath := filepath.Join(userDataPath, fmt.Sprintf("%s.json", title))

	note, err := n.readNoteFile(userDataPath, fmt.Sprintf("%s.json", title))
	if err != nil {
		return fmt.Sprintf("Note '%s' does not exist. Use POST to create it.", title)
	}

	note.Content = content
	note.UpdatedAt = time.Now()

	jsonNote, err := json.Marshal(note)
	if err != nil {
		return fmt.Sprintf("Error marshaling note: %v", err)
	}

	if err := os.WriteFile(filePath, jsonNote, 0644); err != nil {
		return fmt.Sprintf("Error updating note: %v", err)
	}

	return fmt.Sprintf("Note '%s' has been updated successfully.", title)
}

func (n *NoteTool) deleteNote(userDataPath, title string) string {
	filePath := filepath.Join(userDataPath, fmt.Sprintf("%s.json", title))

	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		return fmt.Sprintf("Note '%s' does not exist.", title)
	}

	if err := os.Remove(filePath); err != nil {
		return fmt.Sprintf("Error deleting note: %v", err)
	}

	return fmt.Sprintf("Note '%s' has been deleted successfully.", title)
}

func (n *NoteTool) searchNotes(userDataPath, query string) string {
	files, err := os.ReadDir(userDataPath)
	if err != nil {
		return fmt.Sprintf("Error reading notes directory: %v", err)
	}

	var results []Note
	for _, file := range files {
		if !file.IsDir() {
			note, err := n.readNoteFile(userDataPath, file.Name())
			if err != nil {
				continue
			}
			if strings.Contains(strings.ToLower(note.Title), strings.ToLower(query)) ||
				strings.Contains(strings.ToLower(note.Content), strings.ToLower(query)) {
				results = append(results, note)
			}
		}
	}

	jsonResults, err := json.Marshal(results)
	if err != nil {
		return fmt.Sprintf("Error marshaling search results: %v", err)
	}

	return string(jsonResults)
}

func (n *NoteTool) getNotesByDate(userDataPath, startDate, endDate string) string {
	start, err := time.Parse("2006-01-02", startDate)
	if err != nil {
		return fmt.Sprintf("Invalid start date format. Use YYYY-MM-DD: %v", err)
	}

	end, err := time.Parse("2006-01-02", endDate)
	if err != nil {
		return fmt.Sprintf("Invalid end date format. Use YYYY-MM-DD: %v", err)
	}

	files, err := os.ReadDir(userDataPath)
	if err != nil {
		return fmt.Sprintf("Error reading notes directory: %v", err)
	}

	var results []Note
	for _, file := range files {
		if !file.IsDir() {
			note, err := n.readNoteFile(userDataPath, file.Name())
			if err != nil {
				continue
			}
			if note.CreatedAt.After(start) && note.CreatedAt.Before(end.Add(24*time.Hour)) {
				results = append(results, note)
			}
		}
	}

	jsonResults, err := json.Marshal(results)
	if err != nil {
		return fmt.Sprintf("Error marshaling date results: %v", err)
	}

	return string(jsonResults)
}

File: .teo/skills/notes/scripts/notes.py
import sys
import json
import os
import datetime
import glob

# Constants
DATA_DIR = os.path.join("data", "notes")

def ensure_user_dir(user_id):
    user_path = os.path.join(DATA_DIR, user_id)
    if not os.path.exists(user_path):
        os.makedirs(user_path)
    return user_path

def get_note_path(user_id, title):
    return os.path.join(DATA_DIR, user_id, f"{title}.json")

def validate_input(args):
    action = args.get("action", "").upper()
    if not action:
        return "Error: action is required"
    if action in ["POST", "PUT"]:
        if not args.get("title"):
            return "Error: title is required"
        if not args.get("content"):
            return "Error: content is required"
    if not args.get("user_id"):
        return "Error: user_id is required"
    return None

def get_notes(user_id):
    user_path = ensure_user_dir(user_id)
    files = glob.glob(os.path.join(user_path, "*.json"))
    notes = []
    for file_path in files:
        try:
            with open(file_path, 'r') as f:
                note = json.load(f)
                notes.append(note)
        except Exception:
            continue
    return json.dumps(notes, indent=2)

def get_note_detail(user_id, title):
    file_path = get_note_path(user_id, title)
    if not os.path.exists(file_path):
        return json.dumps({"error": f"Note '{title}' not found"})
    try:
        with open(file_path, 'r') as f:
            note = json.load(f)
        return json.dumps(note, indent=2)
    except Exception as e:
        return json.dumps({"error": str(e)})

def save_note(user_id, title, content):
    ensure_user_dir(user_id)
    file_path = get_note_path(user_id, title)
    if os.path.exists(file_path):
        return json.dumps({"error": f"Note '{title}' already exists. Use PUT to update it."})
    
    note = {
        "title": title,
        "content": content,
        "created_at": datetime.datetime.now().isoformat(),
        "updated_at": datetime.datetime.now().isoformat()
    }
    
    with open(file_path, 'w') as f:
        json.dump(note, f, indent=2)
    return json.dumps({"message": f"Note '{title}' has been saved successfully."})

def update_note(user_id, title, content):
    file_path = get_note_path(user_id, title)
    if not os.path.exists(file_path):
        return json.dumps({"error": f"Note '{title}' does not exist. Use POST to create it."})
    
    with open(file_path, 'r') as f:
        note = json.load(f)
    
    note["content"] = content
    note["updated_at"] = datetime.datetime.now().isoformat()
    
    with open(file_path, 'w') as f:
        json.dump(note, f, indent=2)
    return json.dumps({"message": f"Note '{title}' has been updated successfully."})

def delete_note(user_id, title):
    file_path = get_note_path(user_id, title)
    if not os.path.exists(file_path):
        return json.dumps({"error": f"Note '{title}' does not exist."})
    
    os.remove(file_path)
    return json.dumps({"message": f"Note '{title}' has been deleted successfully."})

def search_notes(user_id, query):
    user_path = ensure_user_dir(user_id)
    files = glob.glob(os.path.join(user_path, "*.json"))
    results = []
    query = query.lower()
    for file_path in files:
        try:
            with open(file_path, 'r') as f:
                note = json.load(f)
                if query in note["title"].lower() or query in note["content"].lower():
                    results.append(note)
        except Exception:
            continue
    return json.dumps(results, indent=2)

def get_notes_by_date(user_id, start_date, end_date):
    user_path = ensure_user_dir(user_id)
    try:
         # Flexible parsing (handling RFC3339 or YYYY-MM-DD)
        start_dt = datetime.datetime.fromisoformat(start_date) if "T" in start_date else datetime.datetime.strptime(start_date, "%Y-%m-%d")
        end_dt = datetime.datetime.fromisoformat(end_date) if "T" in end_date else datetime.datetime.strptime(end_date, "%Y-%m-%d")
        # Ensure end date includes the full day if it was just YYYY-MM-DD
        if "T" not in end_date:
             end_dt = end_dt + datetime.timedelta(days=1) - datetime.timedelta(microseconds=1)

    except ValueError as e:
        return json.dumps({"error": f"Invalid date format: {str(e)}"})
        
    files = glob.glob(os.path.join(user_path, "*.json"))
    results = []
    for file_path in files:
        try:
            with open(file_path, 'r') as f:
                note = json.load(f)
                created_at = datetime.datetime.fromisoformat(note["created_at"])
                # Compare naive or aware datetimes carefully, usually isoformat from python is aware if timezone provided
                # simplified comparison:
                if start_dt.replace(tzinfo=None) <= created_at.replace(tzinfo=None) <= end_dt.replace(tzinfo=None):
                    results.append(note)
        except Exception:
            continue
    return json.dumps(results, indent=2)

def main():
    if len(sys.argv) < 2:
        print(json.dumps({"error": "No arguments provided"}))
        return

    try:
        args = json.loads(sys.argv[1])
    except json.JSONDecodeError:
        print(json.dumps({"error": "Invalid JSON arguments"}))
        return

    error = validate_input(args)
    if error:
        print(json.dumps({"error": error}))
        return

    action = args["action"].upper()
    user_id = args["user_id"]

    if action == "GET":
        print(get_notes(user_id))
    elif action == "GET_DETAIL":
        print(get_note_detail(user_id, args.get("title", "")))
    elif action == "POST":
        print(save_note(user_id, args["title"], args["content"]))
    elif action == "PUT":
        print(update_note(user_id, args["title"], args["content"]))
    elif action == "DELETE":
        print(delete_note(user_id, args.get("title", "")))
    elif action == "SEARCH":
        print(search_notes(user_id, args.get("search", "")))
    elif action == "GET_BY_DATE":
        print(get_notes_by_date(user_id, args.get("start_date", ""), args.get("end_date", "")))
    else:
        print(json.dumps({"error": f"Unknown action: {action}"}))

if __name__ == "__main__":
    main()
